mlflow/tensorflow/__init__.py

```diff
@@ -734,22 +802,30 @@ def predict(self, data):
 
 
 class _KerasModelWrapper:
-    def __init__(self, keras_model, graph, sess):
+    def __init__(self, keras_model, signature):
         self.keras_model = keras_model
-        self._graph = graph
-        self._sess = sess
+        self.signature = signature
 
     def predict(self, data):
-        def _predict(data):
-            if isinstance(data, pandas.DataFrame):
-                predicted = pandas.DataFrame(self.keras_model.predict(data.values))
-                predicted.index = data.index
-            else:
-                predicted = self.keras_model.predict(data)
+        if self.signature is None and isinstance(data, pandas.DataFrame):
+            # This line is for backwards compatibility:
+            # If model signature is not None, when calling
+            # `keras_pyfunc_model.predict(pandas_dataframe)`, `_enforce_schema` will convert
+            # dataframe input into dict input, so in the case `_KerasModelWrapper.predict`
+            # will receive a dict type input.
+            # If model signature is None, `_enforce_schema` can do nothing, and if the input
+            # is dataframe, `_KerasModelWrapper.predict` will receive a dataframe input,
+            # we need to handle this case, to keep backwards compatibility.
+            predicted = pandas.DataFrame(self.keras_model.predict(data.values))
+            predicted.index = data.index
             return predicted
 
-        predicted = _predict(data)
-        return predicted
+        if not isinstance(data, (np.ndarray, list, tuple, dict)):
+            raise MlflowException(
+                f"Unsupported input data type: {type(data)}",
+                INVALID_PARAMETER_VALUE,
+            )
```

Comment:
Do we want to include the list of supported types in the Exception? 
```suggestion
       supported_tf_types = (np.ndarray, list, tuple, dict)
        if not isinstance(data, supported_tf_types):
            raise MlflowException(
                f"Unsupported input data type: {type(data)}. Must be one of: {[x.__name__ for x in supported_tf_types]}",
                INVALID_PARAMETER_VALUE,
            )
```